### PR TITLE
Simplify TranslogWriter to always write to a stream

### PR DESCRIPTION
**Pull request by s1monw**
We used to write into an in-memory buffer and if necessary also allow reading
from the memory buffer if the some translog locations that are not flushed to
the channel need to be read. This commit hides all writing behind a buffered output
stream and if ncecessary flushes all buffered data to the channel for reading. This allows
for several simplifcations like reusing javas build in BufferedOutputStream and removes the
need for read write locks on the translog writer. All thread safety is now achived using
the synchronized primitive.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 1]
```diff
@@ -116,75 +105,59 @@ public Throwable getTragicException() {
...
-    private final void closeWithTragicEvent(Throwable throwable) throws IOException {
-        try (ReleasableLock lock = writeLock.acquire()) {
-            if (tragedy == null) {
-                tragedy = throwable;
-            } else {
-                tragedy.addSuppressed(throwable);
-            }
-            close();
+    private synchronized final void closeWithTragicEvent(Throwable throwable) throws IOException {
+        if (tragedy == null) {
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: Paranoia: can we `assert throwable != null` here?

Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: lol sure

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 2]
```diff
@@ -116,75 +105,59 @@ public Throwable getTragicException() {
...
+            synchronized (this) {
+                ensureOpen(); // this call gives a better exception that the incRef if we are closed by a tragic event
+                channelReference.incRef();
+                try {
+                    final long offsetToSync;
+                    final int opsCounter;
+                    outputStream.flush();
                     offsetToSync = totalOffset;
                     opsCounter = operationCounter;
+                    // we can do this outside of the write lock but we have to protect from
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: Remove this comment (it's stale)?

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 3]
```diff
@@ -116,75 +105,59 @@ public Throwable getTragicException() {
...
+                    // concurrent syncs
+                    ensureOpen(); // just for kicks - the checkpoint happens or not either way
+                    try {
+                        checkpoint(offsetToSync, opsCounter, channelReference);
+                    } catch (Throwable ex) {
+                        closeWithTragicEvent(ex);
+                        throw ex;
+                    }
+                    lastSyncedOffset = offsetToSync;
+                }finally{
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: nit: could you add spaces around brackets?

Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: oh sure

**Comment by mikemccand** [comment]
LGTM, I love this simplification!

**Comment by jpountz** [comment]
:+1:

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 4]
```diff
@@ -116,75 +105,57 @@ public Throwable getTragicException() {
...
-                if (data.length() > buffer.length - bufferCount) {
-                    flush();
-                }
-                data.writeTo(bufferOs);
-                totalOffset += data.length();
-            }
-            operationCounter++;
-            return new Translog.Location(generation, offset, data.length());
+    public synchronized Translog.Location add(BytesReference data) throws IOException {
+        ensureOpen();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: beautiful

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 5]
```diff
@@ -42,6 +42,7 @@
 import java.nio.file.OpenOption;
 import java.nio.file.Path;
 import java.nio.file.StandardOpenOption;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.locks.ReadWriteLock;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: are these needed?

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 6]
```diff
@@ -313,14 +268,9 @@ public boolean syncUpTo(long offset) throws IOException {
...
-                System.arraycopy(buffer, sourcePosition,
-                    targetBuffer.array(), targetBuffer.position(), targetBuffer.limit());
-                targetBuffer.position(targetBuffer.limit());
-                return;
+        if (position+targetBuffer.limit() > getWrittenOffset()) {
+            synchronized (this) {
+                outputStream.flush();
             }
         }
         // we don't have to have a read lock here because we only write ahead to the file, so all writes has been complete
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: nit: read lock -> just lock.

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 7]
```diff
@@ -313,14 +268,9 @@ public boolean syncUpTo(long offset) throws IOException {
...
     protected void readBytes(ByteBuffer targetBuffer, long position) throws IOException {
-        try (ReleasableLock lock = readLock.acquire()) {
-            if (position >= writtenOffset) {
-                assert targetBuffer.hasArray() : "buffer must have array";
-                final int sourcePosition = (int) (position - writtenOffset);
-                System.arraycopy(buffer, sourcePosition,
-                    targetBuffer.array(), targetBuffer.position(), targetBuffer.limit());
-                targetBuffer.position(targetBuffer.limit());
-                return;
+        if (position+targetBuffer.limit() > getWrittenOffset()) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: Nit: strictly speaking i think we need targetBuffer.remaining()  , which is how many bytes we are reading.

**Comment by bleskes** [comment]
awesome. LGTM3

**Pull request merged**